component/authenticator/src/main/java/org/wso2/carbon/identity/application/authenticator/totp/TOTPAuthenticator.java

```diff
@@ -731,4 +834,57 @@ private void setUserClaimValues(AuthenticatedUser authenticatedUser, String user
             throw new AuthenticationFailedException(errorMessage, e);
         }
     }
+
+    private String getFederatedUserId(AuthenticatedUser authenticatedUser, AuthenticationContext context,
+                                      String tenantDomain) throws TOTPException {
+
+        String userId;
+        String username = authenticatedUser.getUserName();
+        String userStoreDomain = authenticatedUser.getUserStoreDomain();
+        String idpName = authenticatedUser.getFederatedIdPName();
+        int tenantId = getTenantID(tenantDomain);
+        int appTenantId = IdentityTenantUtil.getTenantId(context.getTenantDomain());
+
+        try {
+            int idpId = UserSessionStore.getInstance().getIdPId(idpName, appTenantId);
+            userId = UserSessionStore.getInstance().
+                    getUserId(username, tenantId, userStoreDomain, idpId);
+            if (userId == null) {
+                userId = UUID.randomUUID().toString();
+                UserSessionStore.getInstance().storeUserData(userId, username, tenantId,
+                        userStoreDomain, idpId);
+            }
+        } catch (UserSessionException e) {
+            throw new TOTPException("Error while storing session data for user: " + username + " of " +
+                    "user store domain: " + userStoreDomain + " in tenant domain: " + tenantDomain, e);
+        }
+        return userId;
+    }
+
+    private int getTenantID(String tenantDomain) {
+
+        return (tenantDomain == null) ? MultitenantConstants.INVALID_TENANT_ID : IdentityTenantUtil
+                .getTenantId(tenantDomain);
+    }
+
+    private void storeSecretKeyForFedUsers(AuthenticationContext context) throws TOTPException {
+
+        TOTPSecretKeyDAO totpSecretKeyDAO = DAOFactory.getInstance().getTOTPSecretKeyDAO();
+        if (context.getProperty(TOTPAuthenticatorConstants.ENABLE_TOTP) == null || !Boolean
```

Comment:
let's add debug logs for the conditions we check in this method.

Reply:
Fixed